msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java

```diff
@@ -3,76 +3,85 @@
 
 package com.microsoft.aad.msal4j;
 
+import java.nio.charset.StandardCharsets;
+import java.security.Signature;
 import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Date;
+import java.util.Base64;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.UUID;
 
-import com.nimbusds.jose.JWSAlgorithm;
-import com.nimbusds.jose.JWSHeader;
-import com.nimbusds.jose.JWSHeader.Builder;
-import com.nimbusds.jose.crypto.RSASSASigner;
-import com.nimbusds.jose.util.Base64;
-import com.nimbusds.jose.util.Base64URL;
-import com.nimbusds.jwt.JWTClaimsSet;
-import com.nimbusds.jwt.SignedJWT;
-
 final class JwtHelper {
 
     static ClientAssertion buildJwt(String clientId, final ClientCertificate credential,
                                     final String jwtAudience, boolean sendX5c,
                                     boolean useSha1) throws MsalClientException {
-        if (StringHelper.isBlank(clientId)) {
-            throw new IllegalArgumentException("clientId is null or empty");
-        }
-
-        if (credential == null) {
-            throw new IllegalArgumentException("credential is null");
-        }
 
-        final long time = System.currentTimeMillis();
+        ParameterValidationUtils.validateNotBlank("clientId", clientId);
+        ParameterValidationUtils.validateNotNull("credential", clientId);
 
-        final JWTClaimsSet claimsSet = new JWTClaimsSet.Builder()
-                .audience(Collections.singletonList(jwtAudience))
-                .issuer(clientId)
-                .jwtID(UUID.randomUUID().toString())
-                .notBeforeTime(new Date(time))
-                .expirationTime(new Date(time
-                        + Constants.AAD_JWT_TOKEN_LIFETIME_SECONDS
-                        * 1000))
-                .subject(clientId)
-                .build();
-
-        SignedJWT jwt;
         try {
-            JWSHeader.Builder builder = new Builder(JWSAlgorithm.RS256);
+            final long time = System.currentTimeMillis();
+
+            // Build header
+            Map<String, Object> header = new HashMap<>();
+            header.put("alg", "RS256");
```

Comment:
The alg is `PS256` not `RS256` to suggest PSS padding instead of PKCS1 padding. Please have a look at the .NET code for this.

Reply:
The current version of the library uses RS256 here, and has apparently done so since the 'sendX5C' API was first added in 2020 by PR https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/285: 
https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/54f3d44e13e2e3aaa2743d24fd0835fc132db6a0/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java#L49

So this might need a more thorough investigation. Pretty much every integration test sends that "alg=RS256" header when retrieving the token to access ID labs, so the endpoint is either accepting or ignoring it.